pkg/reloader/reloader.go

```diff
@@ -0,0 +1,212 @@
+// It optionally substitutes environment variables in the configuration.
+type Reloader struct {
+	logger           log.Logger
+	promURL          *url.URL
+	cfgFilename      string
+	cfgSubstFilename string
+	ruleDir          string
+	lastCfgHash      []byte
+	lastRulesHash    []byte
+}
+
+// New creates a new reloader that watches the given input file and copies
+// it to the output file path.
+// It optionally substitutes environment variables on copying.
+func New(logger log.Logger, promURL *url.URL, cfgFile, cfgEnvsubstFile, ruleDir string) *Reloader {
+	return &Reloader{
+		logger:           logger,
+		promURL:          promURL,
+		cfgFilename:      cfgFile,
+		cfgSubstFilename: cfgEnvsubstFile,
+	}
+}
+
+// Watch starts to watch the config file and processes it until the context
+// gets canceled.
+func (r *Reloader) Watch(ctx context.Context) error {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return errors.Wrap(err, "create watcher")
+	}
+	defer w.Close()
+
+	cfgDir := filepath.Dir(r.cfgFilename)
+
+	if err := w.Add(cfgDir); err != nil {
+		return errors.Wrap(err, "add file watch")
+	}
+	if _, err := r.applyConfig(); err != nil {
+		return errors.Wrap(err, "initial apply")
+	}
+	if err := r.triggerReload(ctx); err != nil {
+		level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+	}
+
+	tick := time.NewTicker(3 * time.Minute)
+	defer tick.Stop()
+
+	for {
+		select {
+		case event := <-w.Events:
+			level.Debug(r.logger).Log("msg", "received watch event", "op", event.Op, "name", event.Name)
+
+			if event.Name != r.cfgFilename {
+				continue
+			}
+			changes, err := r.applyConfig()
+			if err != nil {
+				level.Error(r.logger).Log("msg", "apply failed", "err", err)
+			}
+			if !changes {
+				continue
+			}
+			if err := r.triggerReload(ctx); err != nil {
+				level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+			}
+
+		case err := <-w.Errors:
+			level.Error(r.logger).Log("msg", "watch error", "err", err)
+
+		case <-tick.C:
+			changes, err := r.refreshRules()
+			if err != nil {
+				level.Error(r.logger).Log("msg", "refreshing rules failed", "err", err)
+			}
+			if !changes {
+				continue
+			}
+			if err := r.triggerReload(ctx); err != nil {
+				level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+			}
+
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func hashFile(h hash.Hash, fn string) error {
+	f, err := os.Open(fn)
+	if err != nil {
+		return err
+	}
+	h.Write([]byte{'\xff'})
+	h.Write([]byte(fn))
+	h.Write([]byte{'\xff'})
+
+	if _, err := io.Copy(h, f); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reloader) applyConfig() (ok bool, err error) {
+	if r.cfgFilename == "" {
+		return false, nil
+	}
+	h := sha256.New()
+
+	if err := hashFile(h, r.cfgFilename); err != nil {
+		return false, errors.Wrap(err, "hash file")
+	}
+	hb := h.Sum(nil)
+
+	if bytes.Equal(r.lastCfgHash, hb) {
```

Comment:
I think we should first substitute the cfg file and then check if it was changed.

Reply:
I assume you expect envvars to never change during runtime... But maybe someone will exec into pod and change them? I guess in that case watch will not trigger event anyways. 

Yea maybe let's stick to yours logic.

Reply:
Can you actually change env vars of a running process? Digging around it a bit it seems to be possible – but then you basically messing around with `gdb` and/or `/proc` already. Doesn't seem like something worth covering.

Reply:
:+1: